packages/happn-3/lib/providers/security-base-auth-provider.js

```diff
@@ -0,0 +1,255 @@
+  }
+
+  get options() {
+    return this.#options;
+  }
+
+  get locks() {
+    return this.#locks;
+  }
+
+  static create(providerConfig, securityFacade, happnConfig) {
+    const providerPathOrInstance =
+      providerConfig.provider != null ? providerConfig.provider : providerConfig;
+
+    // we attach an instance directly to the config
+    if (providerPathOrInstance instanceof SecurityBaseAuthProvider) {
+      return providerPathOrInstance;
+    }
+    const options = providerConfig.options || {};
+    // we attach a factory directly to the config
+    if (providerPathOrInstance instanceof SecurityAuthProviderFactory) {
+      return providerPathOrInstance.create(securityFacade, happnConfig, options);
+    }
+
+    // we are using a filepath
+    let providerClass;
+    try {
+      providerClass = require(providerPathOrInstance);
+    } catch (e) {
+      securityFacade.log.error(
+        `failed to resolve or instantiate auth provider on path: ${providerPathOrInstance}`
+      );
+      throw e;
+    }
+    const providerInst = new providerClass(securityFacade, happnConfig, options);
+    // we are using a filepath that points to a factory
+    if (providerInst instanceof SecurityAuthProviderFactory) {
+      return providerInst.create(securityFacade, happnConfig, options);
+    }
+    return providerInst;
+  }
+
+  accessDenied(errorMessage) {
+    throw this.#securityFacade.error.AccessDeniedError(errorMessage);
+  }
+
+  invalidCredentials(errorMessage) {
+    throw this.#securityFacade.error.InvalidCredentialsError(errorMessage);
+  }
+
+  systemError(errorMessage) {
+    throw this.#securityFacade.error.SystemError(errorMessage);
+  }
+
+  async login(credentials, sessionId, request) {
+    if (credentials.username === '_ANONYMOUS') {
+      if (!this.#config.allowAnonymousAccess) {
+        throw this.#securityFacade.error.InvalidCredentialsError('Anonymous access is disabled');
+      }
+      credentials.password = 'anonymous';
+    }
+    //default is a stateful login
+    if (credentials.type == null) credentials.type = 1;
+
+    if (
+      !((credentials.username && (credentials.password || credentials.digest)) || credentials.token)
+    ) {
+      return this.invalidCredentials('Invalid credentials');
+    }
+    if (
+      !this.#securityFacade.security.checkIPAddressWhitelistPolicy(credentials, sessionId, request)
+    ) {
+      return this.invalidCredentials('Source address access restricted');
+    }
+    if (
+      this.#securityFacade.security.checkDisableDefaultAdminNetworkConnections(credentials, request)
+    ) {
+      return this.accessDenied('use of _ADMIN credentials over the network is disabled');
+    }
+    if (credentials.token) {
+      return this.tokenLogin(credentials, sessionId, request);
+    }
+
+    return this.#userCredsLogin(credentials, sessionId);
+  }
+
+  async tokenLogin(credentials, sessionId, request) {
+    let [authorized, reason] = this.#securityFacade.utils.coerceArray(
```

Comment:
may not need coerceArray - promise returns an array already so assignment of output may just work without it

Reply:
leaving it in for now